left_sidebar: Add option to show left sidebar shortcuts as icons.

Ddharmani3 opened the pull request:
Added a dropdown in display settings having 2 options:
-Text
-Icons

selecting text will keep the left_sidebar layout same as it is, 
while icons will change the layout to only icons in the top horizontal line.

<!-- Describe your pull request here.-->

Fixes: <!-- Issue link, or clear description.-->
#24434

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

![image](https://user-images.githubusercontent.com/64723994/222910324-f8306685-a7c1-4972-bd1a-11521f4417fd.png)

![image](https://user-images.githubusercontent.com/64723994/222911342-6e298f81-3d22-4c91-b7f7-ca1d19939eea.png)

![image](https://user-images.githubusercontent.com/64723994/222910626-d560ebbf-d33e-490f-b07b-333e9766f237.png)

![image](https://user-images.githubusercontent.com/64723994/222910637-893d9ac2-910c-455f-971f-f7b1fac89cb5.png)


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

Ddharmani3 commented:
@alya this PR is ready for review. Could you please take a look and give suggestions?

alya commented:
Thanks! Looking at the screenshots, I think we need an opaque background for the unread count on draft (and starred messages) -- otherwise, the overlap doesn't look good.

Ddharmani3 commented:
@alya I have added opaque background color for the unread count on draft and started messages.
![image](https://user-images.githubusercontent.com/64723994/223538549-7f177d8e-1341-4788-b648-92def9fef84c.png)
![image](https://user-images.githubusercontent.com/64723994/223538528-5dfaedde-5518-432c-abeb-9b26b13a48a3.png)

I will address the three-dot menu issue based on the suggestion from the discussion in [CZO](https://chat.zulip.org/#narrow/stream/101-design/topic/UI.20redesign.3A.20left.20sidebar.20links/near/1495270) as soon as it becomes available.

alya commented:
Thanks!

The three-dot menu should close when a user selects any option, like our other menus. Please update this behavior and post a gif showing the changes.

alya commented:
I don't love how the unread counters look, but it might be good enough just to get a feel for the experience on CZO.

alya commented:
Let's change the setting label to be more specific, as currently it's not clear what "Button labels" refers to. We can try:

"Style for left sidebar links"

Ddharmani3 commented:
> Thanks!
> 
> The three-dot menu should close when a user selects any option, like our other menus. Please update this behavior and post a gif showing the changes.

https://user-images.githubusercontent.com/64723994/226749984-2a20dc57-11f2-49ff-91b5-26ed5b26b21f.mp4

alya commented:
Thanks! I think you can ask @amanagr for a review once the PR passes tests.

Ddharmani3 commented:
@amanagr, all tests have passed and the PR is ready for review now.

amanagr commented:
Pushed to resolve conflicts, reviewing...

amanagr commented on zerver/models.py:
```diff
@@ -1523,6 +1523,11 @@ class UserBaseSettings(models.Model):
+    BUTTON_LABEL_TEXT = 1
+    BUTTON_LABEL_ICON = 2
+    BUTTON_LABEL_CHOICES = [BUTTON_LABEL_TEXT, BUTTON_LABEL_ICON]
+    button_label = models.PositiveSmallIntegerField(default=BUTTON_LABEL_TEXT)
```
`button_label` doesn't sound like a good name for this setting. Perhaps `top_left_sidebar_buttons_style`?

timabbott commented on web/src/drafts.js:
```diff
@@ -241,6 +241,9 @@ function draft_notify() {
     instance.show();
+    if (instance.popper) {
+        instance.popper.querySelector(".tippy-content").style.whiteSpace = "nowrap";
+    }
```
What's the story with this change? It seems like it might deserve its own commit message.

Ddharmani3 replied:
The reason for this change is that in the new layout, the `saved as draft` tooltip was breaking into multiple lines instead of a single horizontal line when a draft was saved. I will create a separate commit to explain the reason for this change and add comments to explain the change as well. 
Before:
![image](https://user-images.githubusercontent.com/64723994/228286923-427324ab-0e43-4daa-ab1b-ce951a9ed5d0.png)
After:
![image](https://user-images.githubusercontent.com/64723994/228286994-cfa33943-7338-4c23-a46a-b8adf21b6596.png)

timabbott commented on web/src/settings_data.ts:
```diff
@@ -1,3 +1,5 @@
+import $ from "jquery";
```
This module is intentionally a data module that does't have any DOM side effects, so we shouldn't import jQuery here.

Ddharmani3 replied:
Fixed.

timabbott commented on web/styles/left_sidebar.css:
```diff
@@ -540,6 +540,72 @@ li.topic-list-item:hover .topic-sidebar-menu-icon {
+    .top_left_drafts .unread_count {
+        background-color: hsl(0, 0.55%, 71.18%) !important;
+        border: 0 !important;
+        color: hsl(0, 0%, 100%) !important;
```
Why do these all need `!important`? We generally try to use selector precedence where possible to ensure the right CSS applies to something.

Ddharmani3 replied:
Removed `!important` and used the right selector precedence to ensure the right CSS applies.

timabbott commented on web/styles/left_sidebar.css:
```diff
@@ -540,6 +540,72 @@ li.topic-list-item:hover .topic-sidebar-menu-icon {
+        display: block;
+        cursor: pointer;
+
+        &i {
```
This CSS syntax is no longer permitted; there's a method in #general for how to have `stylelint` update your PR.

Ddharmani3 replied:
used `.zulip-icon` instead of i

timabbott commented on web/styles/left_sidebar.css:
```diff
@@ -540,6 +540,72 @@ li.topic-list-item:hover .topic-sidebar-menu-icon {
+    display: none;
+}
+
+.icon_layout {
```
This CSS class name is too generic -- there's lots of things in the app that it could potentially refer to. When naming elements, especially those that will have top-level CSS rules like this one, try to pick something unambiguous like `top-left-layout-icons`.

timabbott commented on tools/test-js-with-node:
```diff
@@ -190,6 +192,7 @@ EXEMPT_FILES = make_set(
         "web/src/stream_ui_updates.js",
         "web/src/submessage.js",
         "web/src/subscriber_api.js",
+        "web/src/templates.js",
```
We don't want to add any of these files to this list; but I think after my comments below, the main remaining thing you'll need to do is add a simple `dispatch.test.js` test.

Ddharmani3 replied:
removed `settings_data.js` and `templates.js` from exempt files as they do not have any changes now after addressing all the feedback. I will create new tests for `server_events_dispatch.js` and `top_left_corner.js` in the updated PR as they have lost test coverage due to new changes.

timabbott commented:
@Ddharmani3 thanks for all the work on this! It's a great prototype. I posted a batch of comments in addition to @amanagr's on naming for the database field; I think it's well worth your reading and understanding those since they'll save a lot of time in reviewing your future work. My main high level feedback is to give more thought to names for CSS classes and especially database model names such that they feel globally unique.

(It may also be worth starting a separate thread about names like `top_left_corner` and `global_filter`s being totally inconsistent neither ideal for this area -- there may be a preparatory refactoring PR we should do to make this).

That all said, there's a new idea that Vlad had I think is a much better plan for this part of the app, and in particular doesn't require adding a new setting or a new popover (which were both definitely warts in this design approach). I think it'll still be useful to test-deploy this on chat.zulip.org for a day or so to get some intuition for this shape of UI, though. So my proposal is:
* Remove the setting from this PR (It'll make test-deploying easier, since we skip the database migration).
* Address the simple/tactical feedback above that'll still be relevant with the setting removed (It'll be good practice for you in working on larger projects, but I think most of the feedback above is mostly going to be relevant for you doing future PRs)
* We'll plan to test-deploy this for feedback, and then once Vlad has posted his new ideas, you can pick up work on that plan as a replacement for this PR, since you've already got a bunch of useful experience in working on this part of the codebase.
* Potentially in the meantime, you can start a thread in #frontend on what to call this region and perhaps do some refactoring on the existing names we're using in preparation for that project.

Ddharmani3 commented on web/src/templates.js:
```diff
@@ -127,3 +128,5 @@ Handlebars.registerHelper("tooltip_hotkey_hints", (...hotkeys) => {
     return new Handlebars.SafeString(result);
 });
+
+Handlebars.registerHelper("button_preference", () => get_button_preference());
```
I fixed this by removing the Handlebar Helper and instead passed the layout_prefernce variable to render_left_sidebar() from ui_init.js.

Ddharmani3 commented on web/src/settings_data.ts:
```diff
@@ -198,3 +200,12 @@ export function using_dark_theme(): boolean {
+    }
+    $("#global_filters").removeClass("icon_layout");
+    return "";
+}
```
Created a function that toggles this class inside `top_left_corner.js` and removed any changes made to settings_data.ts including this get_button_preference() and the jquery import [here](https://github.com/zulip/zulip/pull/24582/files/b5d7d8beee388157284e2cf309eb0bcbff5efff6#r1149885537)

Ddharmani3 commented:
> That all said, there's a new idea that Vlad had I think is a much better plan for this part of the app, and in particular doesn't require adding a new setting or a new popover (which were both definitely warts in this design approach). I think it'll still be useful to test-deploy this on chat.zulip.org for a day or so to get some intuition for this shape of UI, though. So my proposal is:
> 
> * Remove the setting from this PR (It'll make test-deploying easier, since we skip the database migration).
> * Address the simple/tactical feedback above that'll still be relevant with the setting removed (It'll be good practice for you in working on larger projects, but I think most of the feedback above is mostly going to be relevant for you doing future PRs)
> * We'll plan to test-deploy this for feedback, and then once Vlad has posted his new ideas, you can pick up work on that plan as a replacement for this PR, since you've already got a bunch of useful experience in working on this part of the codebase.
> * Potentially in the meantime, you can start a thread in #frontend on what to call this region and perhaps do some refactoring on the existing names we're using in preparation for that project.

Thank you, @timabbott, for taking the time to review my PR and provide valuable feedback. As a new open source contributor, I am constantly learning from the community, and your feedback has been especially helpful in improving the readability and concision of my contributions.

I have addressed all the simple/tactical feedback above that'll still be relevant with the setting removed and fixed them in my branch locally.

I have also created a [CZO](https://chat.zulip.org/#narrow/stream/6-frontend/topic/Left.20Sidebar.20Naming.20Refactor/near/1535516) thread for refactoring the naming in the left sidebar as suggested by you. I will create a preparatory refactoring PR once the naming is finalized.

I have made all the necessary changes in frontend and Before proceeding with removing the settings and backend in my PR, I want to confirm that I have correctly understood the proposal described by you above:
From my understanding, Vlad has a better plan for the new UX feature that does not require new settings or a popover.
so before proceeding with that idea we should do a test deploying in CZO for 1 or 2 days for the community feedback on this new feature.
For this in my updated PR, I should remove settings and all the backend to make the test-deploying easier by skipping any new migrations
After test-deployment feedback, I will love to create a replacement pr incorporating the new idea that Vlad will post.

Is my understanding of the proposal correct?
I will push new changes as soon as you confirm the above.

Ddharmani3 commented:
@timabbott I have updated the PR so that it can easily be test-deployed.
This PR keeps the text view by default which can be changed through settings but will be lost on page reload. This is the video of how it works:

https://user-images.githubusercontent.com/64723994/230485261-96ae338c-efa4-437b-a29a-83ae6c05f781.mp4


This PR is ready to be reviewed and test deployed.

Ddharmani3 commented:
@timabbott, I have rebased this PR, do you plan to test-deploy it for CZO feedback or should I close this PR and open a PR for #25902 also asked [here](https://github.com/zulip/zulip/issues/24434#issuecomment-1577814138)?

The behavior is the same as shown in the video in [this comment](https://github.com/zulip/zulip/pull/24582#issuecomment-1499568328).

amanagr commented:
Should this be closed?

Ddharmani3 commented:
> Should this be closed?

Yes, closing in favor of #26035
